### PR TITLE
connection.gd: avoid double UTF-8 encode on every WS send

### PR DESCRIPTION
**Pull request by eth0net**
## Summary

`McpConnection._send_json` runs on every outbound WebSocket frame. It was encoding each payload to UTF-8 **twice**:

1. `text.to_utf8_buffer().size()` — a full encode, used only to measure the byte length for the backpressure check, and
2. `_peer.send_text(text)` — which encodes the same string to UTF-8 again internally to put it on the wire.

The first encode is pure overhead on the hot send path, and it allocates a throwaway `PackedByteArray` the size of the whole payload every time.

## Change

Gate the backpressure check on a cheap upper bound first: UTF-8 uses at most 4 bytes per code point, so `text.length() * 4` upper-bounds the encoded size without touching the bytes. Only when that bound lands near the 4 MiB ceiling do we fall through to an exact `to_utf8_buffer().size()`.

The exact byte count still drives the real overflow decision (`_would_exceed_outbound_backpressure`) and is still what's reported in the backpressure error payload — so **behaviour is unchanged**. The only difference is that a payload comfortably under the limit (i.e. essentially every frame) now encodes once, inside `send_text`, instead of twice.

New static helper `_might_exceed_outbound_backpressure(buffered_bytes, char_count)` mirrors the existing `_would_exceed_outbound_backpressure` and keeps the cheap check unit-testable.

## Why the upper bound is safe

`_might_exceed` is deliberately conservative: because actual UTF-8 bytes can never exceed `char_count * 4`, a "safe" verdict from the cheap gate guarantees the exact check would also pass. It can only ever be *too cautious* (running the exact encode when it turns out not to be needed), never too permissive — so an overflowing payload can never slip past the exact check.

## Tests

`test_project/tests/test_connection.gd`:

- `test_might_exceed_uses_worst_case_four_bytes_per_code_point` — pins the `* 4` worst-case boundary (exactly at the ceiling = safe; one code point over = flagged).
- `test_might_exceed_is_conservative_relative_to_exact_check` — asserts the cheap gate never clears a payload the exact check would reject.

Validated locally on Godot 4.7.1 (headless `--check-only` parse + direct execution of the static helpers).

## Summary by CodeRabbit

* **Performance**
  * Improved outbound WebSocket backpressure checks using a fast upper-bound pre-check to reduce extra work for smaller payloads.
  * Kept precise UTF-8 byte accounting for messages that may approach or exceed the configured outbound buffer limit.

* **Tests**
  * Added unit tests validating the new conservative pre-check behavior across UTF-8 worst-case sizing and consistency with the exact size evaluation.

### CODE REVIEW

File: plugin/addons/godot_ai/connection.gd [hunk 1]
```diff
@@ -570,6 +576,14 @@ static func _would_exceed_outbound_backpressure(buffered_bytes: int, message_byt
 	return buffered_bytes + message_bytes > OUTBOUND_BUFFER_LIMIT_BYTES
 
 
+## Cheap pre-check on the code-point count: UTF-8 uses at most 4 bytes per code
+## point, so `char_count * 4` upper-bounds the encoded size. When even that
+## upper bound fits under the ceiling the payload is definitely safe and we can
+## skip the exact encode; only a positive here warrants `to_utf8_buffer()`.
+static func _might_exceed_outbound_backpressure(buffered_bytes: int, char_count: int) -> bool:
+	return buffered_bytes + char_count * 4 > OUTBOUND_BUFFER_LIMIT_BYTES
```
[the remark this answers is not in the document]
Comment (dsarno) [reply plugin/addons/godot_ai/connection.gd:584]: Declining this one: `char_count` is `String.length()`, and a Godot String stores UTF-32 — so overflowing `* 4` at int64 would need ~2^62 code points, i.e. ~16 EiB of string memory before the multiply could wrap. The 4 MiB backpressure ceiling keeps real payloads ~9 orders of magnitude away. Not adding a guard to the hot path for an unconstructible input.

File: test_project/tests/test_connection.gd [hunk 2]
```diff
@@ -519,6 +519,29 @@ func test_outbound_backpressure_limit_rejects_payload_that_would_overflow() -> v
 	)
 
 
+func test_might_exceed_uses_worst_case_four_bytes_per_code_point() -> void:
+	## The cheap gate must upper-bound the encoded size (<= 4 UTF-8 bytes per
+	## code point) so a "safe" verdict is never wrong.
+	assert_false(McpConnection._might_exceed_outbound_backpressure(0, 1024))
+	var char_budget := McpConnection.OUTBOUND_BUFFER_LIMIT_BYTES / 4
```
[the remark this answers is not in the document]
Comment (dsarno) [reply test_project/tests/test_connection.gd:529]: This is a false positive: in GDScript 2, `int / int` is integer division (yielding an `int`), not a float — and `OUTBOUND_BUFFER_LIMIT_BYTES` (4 MiB) is exactly divisible by 4, so the boundary value is precise. Empirically confirmed: this test just ran green across the full CI matrix. No change needed.

**Comment by dsarno** [comment]
Micro-benchmark for the record (repo practice for GDScript perf changes — macOS, Godot 4.6.3 headless, 20k iterations, representative ~84 KB JSON payload with mixed ASCII + multibyte content):

| gate | per call |
|---|---|
| exact `to_utf8_buffer().size()` (old, every frame) | **75.9 µs** |
| cheap `length() * 4` bound (new, every frame) | **0.025 µs** |

~3000× cheaper on the common path — ~76 µs plus a payload-sized allocation removed from every outbound frame at this size, scaling linearly with payload. Both Copilot findings reviewed and declined with reasoning on the threads (unreachable int64 overflow; integer-division false positive). Content review clean, CI green — LGTM.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: plugin/addons/godot_ai/connection.gd (hunk 1 above)
[the remark this answers is not in the document]
Comment (eth0net) [reply plugin/addons/godot_ai/connection.gd:584]: Agreed on both — thanks for confirming, nothing to change here.

**Review by dsarno — CHANGES_REQUESTED** [review]
The optimization is correct and the safety argument holds: Godot 4 strings are UTF-32 internally, so `String.length()` counts code points, and UTF-8 never exceeds 4 bytes per code point — the cheap gate can only be over-cautious, never permissive. The exact byte count still makes the final overflow decision and still feeds the backpressure error payload, so behavior is unchanged while the common case drops from two encodes to one. The two new tests pin exactly the right properties (the `* 4` boundary and the conservativeness invariant relative to the exact check).

One change requested:

1. **Avoid the untyped integer division in the test** (`test_connection.gd:526`): `McpConnection.OUTBOUND_BUFFER_LIMIT_BYTES / 4` is `int / int`, which trips GDScript's `INTEGER_DIVISION` warning. There's prior art for this pattern in the suite (`test_editor.gd:994`), but new code shouldn't add warning noise. Please either annotate with `@warning_ignore("integer_division")` or write it as an explicit integer op (e.g. `OUTBOUND_BUFFER_LIMIT_BYTES >> 2` or `int(OUTBOUND_BUFFER_LIMIT_BYTES / 4.0)`), keeping the intent (exact quarter of the limit) obvious.

---
_Generated by [Claude Code](https://claude.ai/code)_

**Comment by eth0net** [comment]
Fixed in `fbf3c07` — annotated the `char_budget` line with `@warning_ignore("integer_division")` (with a comment noting the limit is a power of two so the division is exact and intended). Confirmed the `INTEGER_DIVISION` warning is gone via a headless `--check-only` on `test_connection.gd`. Thanks for the benchmark on the record, too.

**Review by dsarno — APPROVED** [review]
fbf3c07 resolves the review point — `@warning_ignore("integer_division")` with a comment explaining the even division. All checks green. Approving.

---
_Generated by [Claude Code](https://claude.ai/code)_
